### ISSUE TITLE
ELM HEFER map pack generation

### ISSUE THREAD

**Issue by CrispinF** [exeGesIS-SDM/ELM#18]
Processing submitted valid SBIs and updating status when completed. Similar to the existing service, i.e. running offline and notifying when complete. Details of what maps to include in the pack to be confirmed, but will need to show the SHINE and NHLE assets clearly against Ordnance Survey mapping at appropriate scales, plus guidance.

Sarah said "we could drop the notification by email, and simply have the download available – given this is a near instant process now rather than a consultation", but I'd note that although there's no consultation period, it won't be near instant... it could take anything from a few minutes to a few hours. This suggests there should either be a notification when ready (to the submitting RPA user?), or perhaps the HEFER download page simply needs to be quite informative, saying things like "No HEFER request submitted for this SBI" or "HEFER for this SBI is being prepared - please try again later", if the download is not yet available.

**Comment by CrispinF** [comment]
@sarahpoppy we need details of what maps should be in the ELLM HEFER pack.

**Comment by CrispinF** [comment]
@sarahpoppy could we have some details here about what maps should be in the ELM HEFER pack? If any of these correlate exactly with maps in the CS HEFER that will be easiest of course (e.g. if the map is the same except for the removal of the HMAA).

**Comment by sarahpoppy** [comment]
@CrispinF Something like the 1:10k HEFER feature maps showing designated and undesignated assets . However the placement of the land parcels on the map tiles is very poor so I would expect that to be improved, also to have legend on the same page.

**Comment by CrispinF** [comment]
We've decided to try to optimise positioning of parcels on the maps by making a new new map grid for each SBI, based on its envelope BBOX.

**Comment by CrispinF** [comment]
Calculate offset of stored national grid to new virtual grid, using minx, miny of the BBOX of the request parcels (+c100m x and y to ensure parcels don't touch map edge on left/bottom); the proceed as before.

**Comment by sarahpoppy** [comment]
Here is a sample for the ELM map output that has been reviewed & agreed with RPA/Defra.   This needs to be a single document rather than a pack to avoid applicants having multiple documents/issues with zip files etc. 

[SFI historic pack example.docx](https://github.com/exeGesIS-SDM/ELM/files/6324181/SFI.historic.pack.example.docx)

**Comment by CrispinF** [comment]
@sarahpoppy by "document" and this example, or you genuinely meaning you want the output in a Microsoft Word docx format?

**Comment by sarahpoppy** [comment]
No, an expecting this to be pdf.  This is just a word document so I could format it and add comments!

**Comment by CrispinF** [comment]
@sarahpoppy this issue also needs a decision on the replacement for "HEFER" (and any other terminology that appears on the CS maps that will be different for ELM).

**Comment by sarahpoppy** [comment]
The wording on the maps should be per the sample provided - which has removed HEFER references.

**Comment by CrispinF** [comment]
@sarahpoppy important & urgent question ... the maps need to tabulate the historic environment features, but those features are not yet in the new database. We're thinking through the possible methods of accessing them.
The simplest involves cloning the tables of HE assets into the new database. For NHLE features this can be a one-off (annual), as the data is updated annually in the CS database. But the SHINE data changes frequently; so it it OK for this SFI pilot to use a snapshot of the SHINE polygons taken now? Or would this have to be updated every day?/week?/month?

**Comment by sarahpoppy** [comment]
@CrispinF The historic environment features need to up-to-date, so the SHINE snapshot needs to be weekly as a minimum please.   The pilot is going to be a pretty intensive period of 4 weeks of people requesting HEFERs, so if it is very resource hungry to build the update mechanism we can rethink this.

**Comment by epennifold** [comment]
To get current SHINE data - new table in SHINE database which mirrors land parcels in ELM - push data across when fetched for SBI and then clear up after. Write view for intersecting HE features with land parcels.

**Comment by joerichardsESDM** [comment]
@sarahpoppy What is your preferred filename format for these packs? I'm using **HEFER-Map-_sbinumberhere_.pdf** as a placeholder for the moment.

**Comment by CrispinF** [comment]
@joerichardsESDM The filenames need to be unique even if an SBI is requested more than once.
So I'd suggest we must incorporate a date/time in format _yyyy_mm_dd_hh_mm_ss.pdf
And, given it's not just a map, it could become:
hefer_pack_{sbinumber}__yyyy_mm_dd_hh_mm_ss.pdf
Would that be OK @sarahpoppy ?

**Comment by sarahpoppy** [comment]
That seems pretty unwieldy to me .   At least lets drop the 'pack' as not needed.  have SFI_HEFER_nnnnnnnnn.pdf where nnn = SBI number

@CrispinF  we discussed that only the most recent pack would be available for download, so do we need to include the yymm when the text is document is date stamped?

**Comment by CrispinF** [comment]
@sarahpoppy the file names for repeat request/downloads really want to be unique, so including the timestamp is sensible and standard practice for output like this. 
So sounds like we can settle on this pattern:
SFI_HEFER_{sbinumber}__yyyy_mm_dd_hh_mm_ss.pdf

**Comment by sarahpoppy** [comment]
Fine!

**Comment by joerichardsESDM** [comment]
I've changed the file naming and also cleaned up some of the code to avoid some edge cases that were throwing errors. Tomorrow I'll start on integrating the designated asset data from SHINE & getting some maps produced.
